### PR TITLE
emcc: IndexError: string index out of range

### PR DESCRIPTION
**Pull request by hydrocat**
Follows on previous pull-request #4759 and on issue #4748
Adds a try-catch to the function is_bitcode on tools/shared.py so that it return a boolean as expected.
The issue was caused by a call to is_bitcode in emcc.py that didn't expect Exceptions.

### CODE REVIEW

**Comment by hydrocat** [comment]
What is the status ?

File: tools/shared.py [hunk 1]
```diff
@@ -1954,17 +1954,21 @@ def is_ar(filename):
...
+      if b[0] == 'B' and b[1] == 'C':
+        return True
+      # look for ar signature
+      elif Building.is_ar(filename):
+        return True
+      # on OS X, there is a 20-byte prefix
+      elif ord(b[0]) == 222 and ord(b[1]) == 192 and ord(b[2]) == 23 and ord(b[3]) == 11:
+        b = open(filename, 'r').read(24)
+        return b[20] == 'B' and b[21] == 'C'
+    except IndexError:
```
Comment (dschuff) [inline tools/shared.py]: Are there any other kinds of errors that we should check here? Can open() or read() throw anything we'd want to catch?

Comment (hydrocat) [inline tools/shared.py]: The function Building.is_ar already logs on errors
maybe we could log fails on the magic signature check and on OSX's prefix check

Comment (hydrocat) [inline tools/shared.py]: Note that the caller function also logs for unknown file type and so we would get logs similar to the following example:
> - Fail magic number check
> - Fail <Building.is_ar's error mesage> 
> - Fail OSX check
>  - Unknown format, not a static library! (From the caller function in emcc.py)

File: tools/shared.py [hunk 2]
```diff
@@ -1954,17 +1954,21 @@ def is_ar(filename):
...
-      return b[20] == 'B' and b[21] == 'C'
+    try:
+      # look for magic signature
+      b = open(filename, 'r').read(4)
+      if b[0] == 'B' and b[1] == 'C':
+        return True
+      # look for ar signature
+      elif Building.is_ar(filename):
+        return True
+      # on OS X, there is a 20-byte prefix
```
Comment (hydrocat) [inline tools/shared.py]: for this, I'm unable to answer

Comment (dschuff) [inline tools/shared.py]: IIRC there is some issue related to the archiver on OSX where it has trouble archiving raw bitcode files (I forget why). So on OSX LLVM (even upstream LLVM) wraps bitcode files in a wrapper header that (IIRC) pads the file out to some particular length.

**Comment by dschuff** [comment]
LGTM, please fix the merge conflict on AUTHORS

**Comment by kripken** [comment]
I think this would already return `False` on small or empty files, as the array accesses would be out of range, so they'd throw and then reach the end return? So this looks good to me.

**Comment by hydrocat** [comment]
Fixed the conflict on the author names

**Comment by kripken** [comment]
Thanks, I think this is ready to merge, however, re-reading it now, I'm not sure why the change from `rb` to `r` when opening those files? We do only compare to characters `B, C` etc., so it seems safe, but it should have worked before too?

**Comment by juj** [comment]
It looks like this commit was incorrectly rebased/merged to latest incoming, dropping old changes. I.e. the PR is now reverting this earlier commit: https://github.com/kripken/emscripten/commit/2d4d7ca12123824f992219e5f4c92a03b9cd6bb7.

Can you rebase the change and drop merge commits from the PR?

**Comment by hydrocat** [comment]
Oops ! I'll look into this. Sorry.

**Comment by kripken** [comment]
Going through old PRs, I noticed this one was approved and just needed to be updated. Updated now, all tests pass, so merging.

Thanks for the PR @hydrocat!

**Comment by hydrocat** [comment]
My pleasure!

**Pull request merged**